rfcs/20200609-logsigmoid/README.md

```diff
@@ -0,0 +1,66 @@
+# Introducing logsigmoid operation
+
+## Introduction
+
+Request comes from DGP graph lib community https://www.dgl.ai. Logsigmoid is
+used in AWS embedding models https://bit.ly/30kWI1f. Currently logsigmoid can be
+implemented with oneDNN by using two eltwise
+(https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html) operations
+logistic and log. Unfortunately current eltwise ops are not fusable.
+Logistic loads data from memory, make computations, stores data and next log
+loads logistic output from memory and stores data (2x loads and 2x stores). While
+it could be possibly 1x load and 1x store if these operations were combined.
+Eltwise is memory bound primitive, so additional loads and store has performance
+impact.
+
+Naming convention taken from: https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html
+FWD derivative calculated using https://www.wolframalpha.com/input/?i=d%2Fdx%28log%281%2F%281%2Bexp%28-x%29%29%29%29
+
+Formulas:
+- forward_logsigmoid - d = log(1/(1+exp(-x)));
```

Comment:
I wonder if there are any expectations on the accuracy of the logsigmoid. We might need to check that with framework guys. I doubt, but it might happen that they consider this complex operation as a whole, that (in theory) could give better accuracy, than chaining it. For instance:
```
log(1/1+e^-x)) = -log(1 + e^-x)
```

If it turns out they use something like this, I doubt we have an option to use just post-ops.

Reply:
Nit:

>```
> log(1/1+e^-x)) = -log(1 + e^-x)
> ```

I think you meant
```
log(1/(1+e^(-x))) = -log(1 + e^(-x))
```

Reply:
fixed, thx!

Reply:
Mentioned formula is used in tensorflow https://www.tensorflow.org/api_docs/python/tf/math/log_sigmoid (For numerical stability, we use y = -tf.nn.softplus(-x)). In other framworks seems that original formula of log_sigmoid is used. Anyway I will ask people who requested feature.

Reply:
Hi ,post-ops may result in numerical unstability, in pytorch also use following trick
```c
    scalar_t max_elem = fmax(0, -*input_data);
    scalar_t z = exp(-max_elem) + exp(-*input_data - max_elem);
    *buffer_data = z;
    *output_data = -(max_elem + log(z));
```
https://github.com/pytorch/pytorch/blob/72e17715dd7e97af518b3e557ded646c23b62e90/aten/src/THNN/generic/LogSigmoid.c#L13

Reply:
So, the conclusions are as follows:
- we are more inclined towards extending list of operations instead of using post-ops;
- w don't want to use `log(1/(1+e^-x))`, because of it may result in numerical unstability;
Open question:
- Which formula we should use for numerical stability tf or pytorch ?

Reply:
That's a good question.
The formula shown by @rongzha1 is protected against overflows, and we do similar trick in (log)softmax. The problem, however, that this approach is no longer an element-wise in some sense -- you need to gather the statistic over the whole buffer. That would prevent the possibility of fusion and will cost us performance.
So, IMO if we implement `logsigmoid`, we implement it straightforward (i.e. TF version)...

@rongzha1, do you think that would be a no-go for PyTorch integration?
In theory, we could have the following work arounds:
1. Use `alpha` parameter for `max_elem`. It won't be computed at run time, but if really required, users will have a know to improve the stability (manually). This is ugly and I really don't think that path forward. But still.
2. Specifically for PyTorch implement `logsigmoid_accurate`, but state in the documentation that this alg_kind cannot be used for fused computations.

Reply:
@piotrchmiel, please add this aspect to the RFC as well.
Also, could you please check what does MxNet use?

Reply:
OK. both are LGTM.  In fact, I'm not working on framework integration.

Reply:
@emfomenk AFAIK, MXNet doesn't have a standalone logsigmoid operator. That means usually it's implemented via `log(sigmoide(x))` in model. Once this primitive is ready in oneDNN library, we can add the corresponding operator to  framework.

Reply:
Thank you, @TaoLv and @romgzha1!

Reply:
@emfomenk Why do you think that we need to gather statistics over the whole buffer when using pytorch approach ? Please correct my if I'm wrong, but based on proposed workaround 1, where you suggested to pass max_elem, as alpha, it suggest that we have one max_elem per entire tensor. But if I didn't missed anything in pytorch code https://github.com/pytorch/pytorch/blob/72e17715dd7e97af518b3e557ded646c23b62e90/aten/src/THNN/generic/LogSigmoid.c#L13 - logsigmoid code is wrapped by macro TH_TENSOR_APPLY3, which is for loop over all elements. So computation `scalar_t max_elem = fmax(0, -*input_data);` is per element not per entire tensor. As it is per element, I don't understand why it is not elementwise in some sense ? It is still single pass over the data.

Reply:
Hi @piotrchmiel,

I don't know which part of the body I used to think when wrote that (but that definitely wasn't brain).
Please disregard my message. You are right, this is not per-tensor but per-element maximum.

I still think we should start with the straightforward formula though :)

Resolving the thread. Approving the RFC.
Thank you and sorry for the noise.